commit 8677c074e2842c0c6bf0763de500924e8573cabd
Author: Seiya Nuta <nuta@seiya.me>
Date:   Sat Jan 14 13:30:24 2023 +0900

    [clang-format] Add AlignTrailingCommentsColumnLimit option

    行の末尾のコメントに日本語 (マルチバイト文字) を使うと、文字の幅を上手く考慮できていないのか、
    コメントの位置が揃わなくなってしまう。

    そこで、AlignTrailingCommentsにおいてコラム幅の制限を無視するオプションを「その場しのぎ」
    の修正として追加する。

    ビルド方法:

    $ git clone https://github.com/llvm/llvm-project.git
    $ cmake -S llvm -B build -G Ninja -DLLVM_ENABLE_PROJECTS=clang -DCMAKE_BUILD_TYPE=Release
    $ cd build
    $ ninja bin/clang-format

diff --git a/clang/include/clang/Format/Format.h b/clang/include/clang/Format/Format.h
index 36472c65a892..a0e136440616 100644
--- a/clang/include/clang/Format/Format.h
+++ b/clang/include/clang/Format/Format.h
@@ -1887,6 +1887,7 @@ struct FormatStyle {
   /// statements unless they contradict other rules.
   /// \version 3.7
   unsigned ColumnLimit;
+  bool DisableColumnLimitInAlignTrailingComments;

   /// A regular expression that describes comments with special meaning,
   /// which should not be split into lines or otherwise changed.
diff --git a/clang/lib/Format/Format.cpp b/clang/lib/Format/Format.cpp
index b09d05799eb4..1eba81a6203e 100644
--- a/clang/lib/Format/Format.cpp
+++ b/clang/lib/Format/Format.cpp
@@ -879,6 +879,8 @@ template <> struct MappingTraits<FormatStyle> {
     IO.mapOptional("BreakInheritanceList", Style.BreakInheritanceList);
     IO.mapOptional("BreakStringLiterals", Style.BreakStringLiterals);
     IO.mapOptional("ColumnLimit", Style.ColumnLimit);
+    IO.mapOptional("DisableColumnLimitInAlignTrailingComments",
+                   Style.DisableColumnLimitInAlignTrailingComments);
     IO.mapOptional("CommentPragmas", Style.CommentPragmas);
     IO.mapOptional("CompactNamespaces", Style.CompactNamespaces);
     IO.mapOptional("ConstructorInitializerIndentWidth",
diff --git a/clang/lib/Format/WhitespaceManager.cpp b/clang/lib/Format/WhitespaceManager.cpp
index 9951906b6af0..55db81be40a6 100644
--- a/clang/lib/Format/WhitespaceManager.cpp
+++ b/clang/lib/Format/WhitespaceManager.cpp
@@ -975,6 +975,8 @@ void WhitespaceManager::alignTrailingComments() {

     if (Style.ColumnLimit == 0)
       ChangeMaxColumn = UINT_MAX;
+    else if (Style.DisableColumnLimitInAlignTrailingComments)
+      ChangeMaxColumn = UINT_MAX;
     else if (Style.ColumnLimit >= Changes[i].TokenLength)
       ChangeMaxColumn = Style.ColumnLimit - Changes[i].TokenLength;
     else
